src/main/scala/is/hail/asm4s/CM.scala

```diff
@@ -1,20 +1,22 @@
-package is.hail.expr
-
-import is.hail.asm4s.{Code, _}
+package is.hail.asm4s
 
 import scala.collection.mutable
 import scala.reflect.ClassTag
 
 object preCM {
-  case class E(m: Map[String, (Type, Code[AnyRef])], fb: Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef])
-  case class S(fa: Array[AnyRef], ec: EvalContext)
+  case class E(m: Map[String, (ClassTag[AnyRef], Code[AnyRef])], fb: Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef])
+  case class S(fa: Array[AnyRef], st: AnyRef)
 
-  def emptyE = E(Map[String, (Type, Code[AnyRef])](), new Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]())
-  def emptyS(ec: EvalContext) = S(Array[AnyRef](), ec)
+  def emptyE = E(Map[String, (ClassTag[AnyRef], Code[AnyRef])](), new Function2Builder[Array[AnyRef],mutable.ArrayBuffer[AnyRef],AnyRef]())
+  def emptyS(ec: AnyRef) = S(Array[AnyRef](), ec)
```

Comment:
I think this needs more radical refactoring.

1. asm4s doens't need the extra binding mechanism for tracking Hail variables, it should support the for-based lexical scoping we talked about in the vectorized numeric operations example.

2.  I'd like to keep asm4s typed unless we're adding an UntypedCode.  I think it is possible with the above constraints.  That mean all this AnyRef nonsense has to go (again, an artifact of the expr language).

3. This should support general function builders.   I'm not 100% sure how to handle typing there.  Maybe FunctionBuilder should be parameterized by the return type of result().